trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/rest/HashCheckInputStream.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.rest;
+
+import com.google.common.hash.Hasher;
+import com.google.common.hash.Hashing;
+import io.airlift.log.Logger;
+import jakarta.ws.rs.WebApplicationException;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import static jakarta.ws.rs.core.Response.Status.UNAUTHORIZED;
+import static java.util.Objects.requireNonNull;
+
+@SuppressWarnings("UnstableApiUsage")
+class HashCheckInputStream
+        extends InputStream
+{
+    private static final Logger log = Logger.get(HashCheckInputStream.class);
+
+    private final InputStream delegate;
+    private final String expectedHash;
+    private final Hasher hasher;
+
+    HashCheckInputStream(InputStream delegate, String expectedHash)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.expectedHash = requireNonNull(expectedHash, "expectedHash is null");
+
+        hasher = Hashing.sha256().newHasher();
+    }
+
+    @Override
+    public int read()
+            throws IOException
+    {
+        int i = delegate.read();
+        if (i < 0) {
+            validateHash();
+            return i;
+        }
+
+        hasher.putByte((byte) (i & 0xff));
+        return i;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len)
+            throws IOException
+    {
+        int result = delegate.read(b, off, len);
+        if (result < 0) {
```

Comment:
My understanding is that `result` would only be `-1` if we fail to read _any_ bytes due to the input stream being exhausted.

If we could read _some_ bytes, which in turn causes us to exhaust the stream, I _think_ this would return `len > result > 0`

Reply:
Just verified

```java
byte[] res = new byte[30];
InputStream x = new ByteArrayInputStream(new byte[] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9});
x.read(res, 0, 100);
```

The final statement returns 10, not -1

Reply:
That is correct, yes. I was wondering whether we should `validateHash` if `result >= 0 && result < len` so that we validate the hash when reading the last chunk

Reply:
I don't think so. Only -1 means the stream has ended.

Reply:
But, maybe we should pass in the expected content length and validate the hash after that many bytes have been read (and fail if too many/few bytes are read)?

Reply:
If callers won't guess that the stream is done based on `result` being less than `len` (which would indicate an EOF has ended, even though we didn't see a -1) then I'm happy with this. wdyt?

Reply:
I changed so that, if available, we use the content length

Reply:
> But, maybe we should pass in the expected content length and validate the hash after that many bytes have been read (and fail if too many/few bytes are read)?

This comment didn't show on my side until recently, this sounds like a good idea!